File: 백준/Silver/2468. 안전 영역/안전 영역.py
from collections import deque
dx = [0,0,-1,1]
dy = [-1,1,0,0]
def bfs(x,y,t):
  q = deque()
  q.append((x,y))
  check[x][y] = True
  while q:
    x,y = q.popleft()
    for k in range(4):
      nx = x + dx[k]
      ny = y + dy[k]
      if (0 <= nx < n and 0 <= ny < n) and not check[nx][ny]:
        if a[nx][ny] > t:
          check[nx][ny] = True
          q.append((nx,ny))
n = int(input())
a = [list(map(int,input().split())) for _ in range(n)]
m = max([max(i) for i in a])
res = 1
for i in range(m):
  cnt = 0
  check = [[False]*n for _ in range(n)]
  for j in range(n):
    for k in range(n):
      if a[j][k] > i and not check[j][k]:
        bfs(j,k,i)
        cnt += 1
  res = max(res,cnt)
print(res)

File: 백준/Silver/2468. 안전 영역/안전 영역.cc
#include <iostream>
#include <algorithm>
#include <queue>
#include <cstring>
using namespace std;

int n;
int map[101][101];
int dx[4] = { 0,0,-1,1 };
int dy[4] = { -1,1,0,0 };
int res = 1;
bool check[101][101] = { false };

void bfs(int x, int y, int temp) {
    queue<pair<int, int>>q;
    q.push({ x,y });
    check[x][y] = true;
    while (!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();

        for (int k = 0; k < 4; k++) {
            int nx = x + dx[k];
            int ny = y + dy[k];

            if (nx < n && nx >= 0 && ny < n && ny >= 0 && !check[nx][ny]) {
                if (map[nx][ny] > temp) {
                    check[nx][ny] = true;
                    q.push({ nx,ny });
                }
            }
        }
    }
}
int main() {
    cin >> n;
    int a = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> map[i][j];
            a = max(map[i][j], a);
        }
    }
    for (int k = 0; k <= a; k++) {
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (map[i][j] > k && !check[i][j]) {
                    bfs(i, j, k);
                    cnt++;
                }
            }
            res = max(cnt, res);
        }
        memset(check, false, sizeof(check));
    }
    cout << res;
    return 0;
}